tokio/src/runtime/thread_pool/queue.rs

```diff
@@ -0,0 +1,550 @@
+            let batch_tail = batch_tail.into_raw();
+
+            debug_assert!(get_next(batch_tail).is_none());
+
+            let mut p = self.pointers.lock().unwrap();
+
+            if let Some(tail) = p.tail {
+                set_next(tail, Some(batch_head));
+            } else {
+                p.head = Some(batch_head);
+            }
+
+            p.tail = Some(batch_tail);
+
+            // Increment the count.
+            //
+            // All updates to the len atomic are guarded by the mutex. As such,
+            // a non-atomic load followed by a store is safe.
+            let len = self.len.unsync_load();
+
+            self.len.store(len + num, Release);
+        }
+    }
+
+    pub(super) fn pop(&self) -> Option<task::Notified<T>> {
+        // Fast path, if len == 0, then there are no values
+        if self.is_empty() {
+            return None;
+        }
+
+        unsafe {
+            let mut p = self.pointers.lock().unwrap();
+
+            // It is possible to hit null here if another thread poped the last
+            // task between us checking `len` and acquiring the lock.
+            let task = p.head?;
+
+            p.head = get_next(task);
+
+            if p.head.is_none() {
+                p.tail = None;
+            }
+
+            set_next(task, None);
+
+            // Decrement the count.
+            //
+            // All updates to the len atomic are guarded by the mutex. As such,
+            // a non-atomic load followed by a store is safe.
+            self.len.store(self.len.unsync_load() - 1, Release);
+
+            Some(task::Notified::from_raw(task))
+        }
+    }
+}
+
+impl<T: task::ScheduleSendOnly> Inject<T> {
+    /// Pushes a value into the queue.
+    pub(super) fn push(&self, task: task::Notified<T>) {
+        unsafe {
+            // Acquire queue lock
+            let mut p = self.pointers.lock().unwrap();
+
+            if p.is_closed {
+                // Drop the mutex to avoid a potential deadlock when
+                // re-entering.
+                drop(p);
+                drop(task);
+                return;
+            }
+
+            // Check if the queue is closed. This must happen in the lock.
+            let len = self.len.unsync_load();
+            let task = task.into_raw();
+
+            // The next pointer should already be null
+            debug_assert!(get_next(task).is_none());
+
+            if let Some(tail) = p.tail {
+                set_next(tail, Some(task));
+            } else {
+                p.head = Some(task);
+            }
+
+            p.tail = Some(task);
+
+            self.len.store(len + 1, Release);
+        }
+    }
+}
+
+impl<T: 'static> Drop for Inject<T> {
+    fn drop(&mut self) {
+        if !std::thread::panicking() {
+            assert!(self.pop().is_none(), "queue not empty");
+        }
+    }
+}
+
+fn get_next(header: NonNull<task::Header>) -> Option<NonNull<task::Header>> {
```

Comment:
This and the `fn` below should both be `unsafe fn`.

Reply:
As discussed, for now, private fns in unsafe **modules** aren't being annotated with `unsafe`.